transformer_engine/pytorch/module/linear.py

```diff
@@ -695,14 +695,20 @@ def backward(ctx, grad_output: torch.Tensor) -> Tuple[Union[torch.Tensor, None],
                     # all-gather with wgrad GEMM. Also, we can't
                     # convert row-scaled MXFP8 to column-scaled, so we
                     # can't reuse the grad output that was gathered
-                    # for the dgrad GEMM. We work around with blocking
-                    # all-gather for column-scaled MXFP8 data.
+                    # for the dgrad GEMM. We work around by explicitly
+                    # overlapping the NCCL operation with the dgrad GEMM.
                     ctx.grad_output_quantizer.set_usage(rowwise=False, columnwise=True)
-                    grad_output, _ = gather_along_first_dim(
-                        grad_output_arg,
-                        ctx.tp_group,
-                        quantizer=ctx.grad_output_quantizer,
-                    )
+                    s = ub_obj_dgrad.get_communication_stream()
+                    with torch.cuda.stream(s):
+                        grad_output, grad_output_work = gather_along_first_dim(
+                            grad_output_arg,
+                            ctx.tp_group,
+                            async_op=True,
+                            quantizer=ctx.grad_output_quantizer,
+                        )
```

Comment:
Since `async_op` uses a pre-allocated torch stream assigned for the TP communication group, grepping the UB communication stream and using it as a context is pretty much a no-op.

Also, the async communication in `gather_along_first_dim` would do synchronization with torh main stream, right? Then, we cannot pull the communication start timeline before the last GEMM chunk.

Reply:
> grepping the UB communication stream and using it as a context is pretty much a no-op.
Ok maybe I am misunderstanding, but my understanding is the point of this is that the async_op creates an cudaEvent dependency between the current stream and the preallocated stream. By using `with` block it changes which stream it syncs with. As you say without this it would sync with the main stream instead and not overlap at all.

Reply:
The stream synchronization logic seems correct to me. We could accomplish the same thing with:
```python
with torch.cuda.stream(s):
    grad_output, _ = gather_along_first_dim(grad_output_arg, ctx.tp_group, quantizer=ctx.grad_output_quantizer)
torch.cuda.current_stream().wait_stream(s)
```

We are effectively serializing the UB and NCCL communication, which shouldn't have an effect if we are fully utilizing bandwidth. This also avoids some correctness issues I saw in https://github.com/NVIDIA/TransformerEngine/pull/1711 when I had simultaneous UB and NCCL communication (see https://github.com/NVIDIA/TransformerEngine/pull/1711/commits/8e63f8cc21e8daf937ad84b80388dcea7f136650).

Reply:
> We could accomplish the same thing with:
> ```with torch.cuda.stream(s):
>    grad_output, _ = gather_along_first_dim(grad_output_arg, ctx.tp_group, quantizer=ctx.grad_output_quantizer)
> torch.cuda.current_stream().wait_stream(s)```

@timmoon10 @erhoo82 do you have a preference. Technically, the current version has one less cuda event synchronize but that should not impact perf so we can pick whichever you think is more readable.
The only real advantage of the current one is we can easily get more overlap if we add any kernels between the AG and the GEMM